sdks/java/core/src/main/java/org/apache/beam/sdk/util/UnboundedScheduledExecutorService.java

```diff
@@ -201,9 +201,26 @@ public UnboundedScheduledExecutorService() {
         new ThreadPoolExecutor(
             0,
             Integer.MAX_VALUE, // Allow an unlimited number of re-usable threads.
-            Long.MAX_VALUE,
-            TimeUnit.NANOSECONDS, // Keep non-core threads alive forever.
-            new SynchronousQueue<>(),
+            // Put a high-timeout on non-core threads. This reduces memory for per-thread caches
+            // over time.
+            1,
+            TimeUnit.HOURS,
+            new SynchronousQueue<Runnable>() {
+              @Override
+              public boolean offer(Runnable r) {
+                try {
+                  // By blocking for a little we hope to delay thread creation if there are existing
```

Comment:
This is a good short term solution for tasks to be more likely paired up with existing threads, if the system is very busy then I'm not sure how much this will help.

I had originally thought of storing the set of unused and available threads instead of using this queue over tasks but there is higher communication overhead to make that happen since you insert threads when they are done, pull them out when you want to assign something and then have to notify them to wake them up.

Reply:
Ack